osu.Game/Screens/Play/Player.cs

```diff
@@ -542,7 +542,11 @@ protected void PerformExit(bool showDialogFirst)
 
                 // if the score is ready for display but results screen has not been pushed yet (e.g. storyboard is still playing beyond gameplay), then transition to results screen instead of exiting.
                 if (prepareScoreForDisplayTask != null)
+                {
+                    completionProgressDelegate?.Cancel();
+                    completionProgressDelegate = null;
                     updateCompletionState(true);
+                }
```

Comment:
This shouldn't cancel and redo the whole completion state again, you would instead want this logic to not be invoked when `completionProgressDelegate` is already invoked.

As this logic is meant to update the completion if: 
 1. the player wants to exit, 
 2. the score is being prepared, 
 3. but the `completionProgressDelegate` has not been scheduled yet.

Reply:
I could change it to `if(prepareScoreForDisplayTask != null && completionProgressDelegate == null)`, but this will make the game transition to the song select screen instead of showing the score immediately. If that's fine I can make the change, I'll also update the test.

I did try just returning if the delegate isn't null, but this makes the game stuck because the delegate is cancelled at the beginning of `PerformExit()`. Is there a reason why it isn't set to null afterwards?

https://github.com/ppy/osu/blob/495eb04e2fc0a453ed63f56fda3d9afd1efb630c/osu.Game/Screens/Play/Player.cs#L510-L511

Reply:
> I could change it to if(prepareScoreForDisplayTask != null && completionProgressDelegate == null), but this will make the game transition to the song select screen instead of showing the score immediately. If that's fine I can make the change, I'll also update the test.

That's how osu!lazer behaves prior to the storyboard outro PR, which this should match to for now (i.e. shouldn't change behaviour on a regression-fixing PR), so yes.

Reply:
This cancellation is redundant. Please check the very first line of this method, which runs unconditionally.

https://github.com/ppy/osu/blob/495eb04e2fc0a453ed63f56fda3d9afd1efb630c/osu.Game/Screens/Play/Player.cs#L508-L511

If there is a bug here, then it is that the delegate field is not cleared after cancellation.